File: CASHIQ-CASHMANAGEMENT-REACT-FRONTEND/src/services/DebtService.ts
import api from '../utils/AxiosConfig';
import type { DebtEntryDTO, DebtSummaryDTO } from '../models/Debt';

const API_URL = 'http://localhost:8080/api/debts';

/**
 * Service for handling Debt and IOU API interactions.
 */
const DebtService = {

    /**
     * Fetches the summary of debts (Total Owed vs Owed To) for a specific user.
     * @param userId - The ID of the user.
     * @returns Promise resolving to DebtSummaryDTO.
     */
    getDebtSummary: async (userId: number): Promise<DebtSummaryDTO> => {
        try {
            const response = await api.get(`${API_URL}/summary`, {
                params: { userId }
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching debt summary:", error);
            throw error;
        }
    },

    /**
     * Fetches all debt records for a user.
     * @param userId - The ID of the user.
     * @returns Promise resolving to an array of DebtEntryDTO.
     */
    getAllDebts: async (userId: number): Promise<DebtEntryDTO[]> => {
        try {
            const response = await api.get(API_URL, {
                params: { userId }
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching all debts:", error);
            throw error;
        }
    },

    /**
     * Creates a new debt or IOU record.
     * @param userId - The ID of the user.
     * @param debtData - The debt object to create.
     * @returns Promise resolving to the created DebtEntryDTO.
     */
    createDebt: async (userId: number, debtData: DebtEntryDTO): Promise<DebtEntryDTO> => {
        try {
            const response = await api.post(API_URL, debtData, {
                params: { userId }
            });
            return response.data;
        } catch (error) {
            console.error("Error creating debt:", error);
            throw error;
        }
    },

    /**
     * Records a payment towards a debt.
     * @param userId - The ID of the user.
     * @param debtId - The ID of the debt to pay.
     * @param amount - The amount to pay.
     * @returns Promise resolving to the updated DebtEntryDTO.
     */
    recordPayment: async (userId: number, debtId: number, amount: number): Promise<DebtEntryDTO> => {
        try {
            const response = await api.put(`${API_URL}/${debtId}/payment`, { amount }, {
                params: { userId }
            });
            return response.data;
        } catch (error) {
            console.error("Error recording payment:", error);
            throw error;
        }
    },

    /**
     * Deletes a debt record.
     * @param userId - The ID of the user.
     * @param debtId - The ID of the debt to delete.
     * @returns Promise resolving when deletion is complete.
     */
    deleteDebt: async (userId: number, debtId: number): Promise<void> => {
        try {
            await api.delete(`${API_URL}/${debtId}`, {
                params: { userId }
            });
        } catch (error) {
            console.error("Error deleting debt:", error);
            throw error;
        }
    }
};

export default DebtService;

File: CASHIQ-CASHMANAGEMENT-MICROSERVICES/CASHIQ-CASHMANAGEMENT-MANAGE-CORE/src/main/java/com/cashiq/cashmanagement/services/DebtService.java
package com.cashiq.cashmanagement.services;

import com.cashiq.cashmanagement.dto.DebtEntryDTO;
import com.cashiq.cashmanagement.dto.DebtSummaryDTO;
import org.springframework.http.ResponseEntity;

import java.math.BigDecimal;
import java.util.List;

/**
 * Service interface for managing Debt and IOU logic.
 */
public interface DebtService {

    /**
     * Retrieves all debts for a given user.
     *
     * @param userId ID of the user.
     * @return List of DebtEntryDTOs.
     */
    List<DebtEntryDTO> getAllDebts(Long userId);

    /**
     * Retrieves a summary of total debts owed by and to the user.
     *
     * @param userId ID of the user.
     * @return DebtSummaryDTO containing totals.
     */
    DebtSummaryDTO getDebtSummary(Long userId);

    /**
     * Creates a new debt record.
     *
     * @param userId       ID of the user.
     * @param debtEntryDTO Data for the new debt.
     * @return The created DebtEntryDTO.
     */
    DebtEntryDTO createDebt(Long userId, DebtEntryDTO debtEntryDTO);

    /**
     * Records a payment towards a specific debt, reducing the balance.
     *
     * @param userId ID of the user (for verification).
     * @param debtId ID of the debt to pay.
     * @param amount Amount to pay/reduce.
     * @return The updated DebtEntryDTO.
     */
    DebtEntryDTO recordPayment(Long userId, Long debtId, BigDecimal amount);

    /**
     * Deletes a debt record.
     *
     * @param userId ID of the user (for verification).
     * @param debtId ID of the debt to delete.
     * @return Response status.
     */
    ResponseEntity<?> deleteDebt(Long userId, Long debtId);
}
